rpc/namenode.go

```diff
@@ -310,21 +318,38 @@ func (c *NamenodeConnection) readResponse(method string, resp proto.Message) err
 // +-----------------------------------------------------------+
 // |  Auth protocol, 1 byte (Auth method None = 0x00)          |
 // +-----------------------------------------------------------+
+
+// +-----------------------------------------------------------+
+// | Authentication handshake happens here, if required.       |
+// +-----------------------------------------------------------+
+
+// This is the request content.
+// +-----------------------------------------------------------+
 // |  uint32 length of the next two parts                      |
 // +-----------------------------------------------------------+
 // |  varint length + RpcRequestHeaderProto                    |
 // +-----------------------------------------------------------+
 // |  varint length + IpcConnectionContextProto                |
 // +-----------------------------------------------------------+
 func (c *NamenodeConnection) writeNamenodeHandshake() error {
+	if c.kerberosClient != nil {
+		return c.writeNamenodeKerberosHandshake()
+	}
+	return c.writeNamenodeNoAuthHandshake()
+}
+
+func (c *NamenodeConnection) writeNamenodeNoAuthHandshake() error {
```

Comment:
Doesn't really seem necessary to make this two separate functions. There's only two places where it branches that I see: choosing which auth protocol flag to include and the call to `doKerberosHandshake`.